File: built-in plugins dsp/mb-comp-transient.cpp
/**
 * MB Transient Shaper
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Attack and sustain shaper
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_TRANSIENT_H
#define MB_COMP_TRANSIENT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompTransient : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-transient";
    static constexpr const char* PLUGIN_NAME    = "MB Transient Shaper";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float attack = 0f;  // range [-100, 100]
    float sustain = 0f;  // range [-100, 100]
    };

    MbCompTransient() = default;
    ~MbCompTransient() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.attack = std::clamp(params.attack, -100f, 100f);
        params.sustain = std::clamp(params.sustain, -100f, 100f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Transient Shaper
        return input;
    }
};

#endif // MB_COMP_TRANSIENT_H

File: built-in plugins dsp/mb-comp-transient.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCompTransientPlugin: PluginDefinition = { id: 'mb-comp-transient', slug: 'mb-comp-transient', name: 'MB Transient Shaper', category: 'effect', type: 'compressor', version: '1.0.0', description: 'Attack and sustain shaper', author: 'Max Booster', grade: 'A', parameters: [{ id: 'attack', name: 'Attack', type: 'float', defaultValue: 0, minValue: -100, maxValue: 100, automatable: true }, { id: 'sustain', name: 'Sustain', type: 'float', defaultValue: 0, minValue: -100, maxValue: 100, automatable: true }], defaultPreset: { attack: 0, sustain: 0 } };

export default MbCompTransientPlugin;
